recipes/libnvshmem/meta.yaml

```
@@ -0,0 +1,203 @@
+      - lib/nvshmem_transport*.so.*
+    requirements:
+      build:
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
+        - {{ compiler('cuda') }}
+        - arm-variant * {{ arm_variant_type }}  # [aarch64]
+        - {{ stdlib("c") }}
+      host:
+        - cuda-version {{ cuda_compiler_version }}
+        - cuda-cudart-dev   # [(cuda_compiler_version or "").startswith("12")]
+        - cudatoolkit       # [(cuda_compiler_version or "").startswith("11")]
+        - ucx
+      run:
+        - {{ pin_compatible("cuda-version", min_pin="x", max_pin="x") }}
+        - libpciaccess
+        - libpmix
+        - nccl
+        - pmix-bin
+        - ucx
+      run_constrained:
+        - arm-variant * {{ arm_variant_type }}  # [aarch64]
+    test:
+      commands:
+        - test -L $PREFIX/lib/libnvshmem_host.so.{{ soname }}
+        - test -f $PREFIX/lib/libnvshmem_host.so.{{ version }}
+        - test -L $PREFIX/lib/nvshmem_bootstrap_mpi.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_bootstrap_mpi.so.{{ soname }}.0.0
+        - test -L $PREFIX/lib/nvshmem_bootstrap_pmi.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_bootstrap_pmi.so.{{ soname }}.0.0
+        - test -L $PREFIX/lib/nvshmem_bootstrap_pmi2.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_bootstrap_pmi2.so.{{ soname }}.0.0
+        - test -L $PREFIX/lib/nvshmem_bootstrap_pmix.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_bootstrap_pmix.so.{{ soname }}.0.0
+        - test -L $PREFIX/lib/nvshmem_bootstrap_shmem.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_bootstrap_shmem.so.{{ soname }}.0.0
+        - test -L $PREFIX/lib/nvshmem_bootstrap_uid.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_bootstrap_uid.so.{{ soname }}.0.0
+        - test -L $PREFIX/lib/nvshmem_transport_ibdevx.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_transport_ibdevx.so.{{ soname }}.0.0
+        - test -L $PREFIX/lib/nvshmem_transport_ibgda.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_transport_ibgda.so.{{ soname }}.0.0
+        - test -L $PREFIX/lib/nvshmem_transport_ibrc.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_transport_ibrc.so.{{ soname }}.0.0
+        - test -L $PREFIX/lib/nvshmem_transport_libfabric.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_transport_libfabric.so.{{ soname }}.0.0
+        - test -L $PREFIX/lib/nvshmem_transport_ucx.so.{{ soname }}
+        - test -f $PREFIX/lib/nvshmem_transport_ucx.so.{{ soname }}.0.0
+
+  - name: libnvshmem-dev
+    build:
+      missing_dso_whitelist:
+        - "*libcuda.so*" # driver
+        - "*libnvshmem_host.so.{{ soname }}" # avoids: ERROR (libnvshmem-dev,bin/perftest/device/pt-to-pt/shmem_put_bw): lib/libnvshmem_host.so.3 not found in any packages. We DO test for its existence though.
+        - "*libmpi*"
+        - "*libpmix*"
+        - "*liboshmem*"
+    run_exports:
+      - {{ pin_subpackage("libnvshmem" ~ soname, max_pin="x") }}
```

```suggestion
      run_exports:
        - {{ pin_subpackage("libnvshmem" ~ soname, max_pin=None) }}
```
I think this needs to be indented one level?

Also, the max_pin should be None if the project actually uses the soname to indicate ABI compatability.

---

Fixed indentation and set `max_pin` to `None`. But I wonder how we can enforce ABI compatibility strictly ? Are leaving the responsibility to the end user to understand what they are doing based on the `soname` ?

---

If we are building libnvshmem version 1.2.3 then this template would be de-templated to something like:

```
- libnvshmem1 >=1.2.3
```

The `soname` is already a promise of ABI compatibility because of the way that unix searches for shared libraries. Any library named `foo.so.4` is assumed to be backward compatible with any earlier `foo.so.4`. Therefore, any library looking for libnvshmem5 can run with the latest version of whatever libnvshmem5 is available. This is why there is no upper bound.

The `libnvshmem3` package is not intended to be installed directly by humans. All downstream users should be installing `libnvshmem-dev`. If you think end users of this package want to be able to install the shared libraries directly without any of the headers or config, we can add a non-versioned `libnvshmem` output.

---

Thanks for the explanation. This helps!